File: fixtures/scale/java/medium/src/calc/repl/Repl.java
package calc.repl;

import java.io.InputStream;
import java.io.PrintStream;

import calc.app.Calculator;
import calc.errors.CalcException;
import calc.io.ResultPrinter;

public class Repl {
    private final Session session;
    private final CommandHandler commands;

    public Repl(Session session) {
        this.session = session;
        this.commands = new CommandHandler(session);
    }

    public int loop(InputStream in, PrintStream out) {
        InputReader reader = new InputReader(in);
        Calculator calculator = new Calculator(session);
        ResultPrinter printer = new ResultPrinter(session.getSettings());
        String line;
        while ((line = reader.nextLine()) != null) {
            if (line.isBlank()) continue;
            if (line.startsWith(":")) {
                if (!commands.handle(line, out)) return 0;
                continue;
            }
            try {
                double value = calculator.evaluateLine(line);
                session.getHistory().record(line, value);
                printer.print(out, value);
            } catch (CalcException e) {
                out.println("error: " + e.getMessage());
            }
        }
        return 0;
    }
}

File: fixtures/scale/dart/medium/lib/repl/repl.dart
import 'dart:io';

import '../app/calculator.dart';
import '../errors/calc_exception.dart';
import '../io/result_printer.dart';
import 'command_handler.dart';
import 'session.dart';

class Repl {
  final Session session;
  late final CommandHandler commands;

  Repl(this.session) {
    commands = CommandHandler(session);
  }

  int loop() {
    final calculator = Calculator(session);
    final printer = ResultPrinter(session.settings);
    String? line;
    while ((line = stdin.readLineSync()) != null) {
      final text = line!.trim();
      if (text.isEmpty) continue;
      if (text.startsWith(':')) {
        if (!commands.handle(text)) return 0;
        continue;
      }
      try {
        final value = calculator.evaluateLine(text);
        session.history.record(text, value);
        printer.printValue(value);
      } on CalcException catch (e) {
        print('error: ${e.message}');
      }
    }
    return 0;
  }
}
